NPE on write error

### Specifications

* Client Version: 0.8.0
* InfluxDB Version: InfluxDB Edge (snapshot)
* Platform: MacOS


### Code sample to reproduce problem

```java
influxDBClient.writePoints(points);
```


### Expected behavior

Fail with a descriptive exception

### Actual behavior

Fail with an NPE

### Additional info

Body from HTTP response looks like:
```
{"error":"parsing failed for write_lp endpoint","data":{"original_line":"sample sample_interval=30,value=0 1715274737218000000","line_number":1,"error_message":"invalid field value in line protocol for field 'sample_interval' on line 0: expected type iox::column_type::field::integer, but got iox::column_type::field::float"}}
```

Which triggers this exception:
```
java.util.concurrent.CompletionException: java.lang.NullPointerException: Cannot invoke "com.fasterxml.jackson.databind.JsonNode.asText()" because the return value of "com.fasterxml.jackson.databind.JsonNode.get(String)" is null

	at java.base/java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:315)
	at java.base/java.util.concurrent.CompletableFuture.completeThrowable(CompletableFuture.java:320)
	at java.base/java.util.concurrent.CompletableFuture$AsyncRun.run$$$capture(CompletableFuture.java:1807)
	at java.base/java.util.concurrent.CompletableFuture$AsyncRun.run(CompletableFuture.java)
	at java.base/java.util.concurrent.ThreadPerTaskExecutor$TaskRunner.run(ThreadPerTaskExecutor.java:314)
	at java.base/java.lang.VirtualThread.run(VirtualThread.java:309)
Caused by: java.lang.NullPointerException: Cannot invoke "com.fasterxml.jackson.databind.JsonNode.asText()" because the return value of "com.fasterxml.jackson.databind.JsonNode.get(String)" is null
	at com.influxdb.v3.client.internal.RestClient.request(RestClient.java:179)
	at com.influxdb.v3.client.internal.InfluxDBClientImpl.writeData(InfluxDBClientImpl.java:323)
	at com.influxdb.v3.client.internal.InfluxDBClientImpl.writePoints(InfluxDBClientImpl.java:157)
	at com.influxdb.v3.client.internal.InfluxDBClientImpl.writePoints(InfluxDBClientImpl.java:152)
	at com.influxdb.v3.client.InfluxClientFactory_ProducerMethod_newClient_Ux0i73pEPhnEmmdrKxjBuioA9j8_ClientProxy.writePoints(Unknown Source)
```

Relevant code block: https://github.com/InfluxCommunity/influxdb3-java/blob/v0.8.0/src/main/java/com/influxdb/v3/client/internal/RestClient.java#L179

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -1,5 +1,9 @@
 ## 0.9.0 [unreleased]
 
+### Bug Fixes
+
+1. [#148](https://github.com/InfluxCommunity/influxdb3-java/pull/148): InfluxDB Edge (OSS) error handling
+
 ## 0.8.0 [2024-06-24]
 
 ### Features

--- a/src/main/java/com/influxdb/v3/client/internal/RestClient.java
+++ b/src/main/java/com/influxdb/v3/client/internal/RestClient.java
@@ -176,7 +176,12 @@ final class RestClient implements AutoCloseable {
             String body = response.body();
             if (!body.isEmpty()) {
                 try {
-                    reason = objectMapper.readTree(body).get("message").asText();
+                    final var obj = objectMapper.readTree(body);
+                    if (obj.hasNonNull("message")) { // Cloud
+                        reason = obj.get("message").asText();
+                    } else if (obj.hasNonNull("error_message")) { // OSS (Edge)
+                        reason = obj.get("error_message").asText();
+                    }
                 } catch (JsonProcessingException e) {
                     LOG.debug("Can't parse msg from response {}", response);
                 }

--- a/src/main/java/com/influxdb/v3/client/internal/RestClient.java
+++ b/src/main/java/com/influxdb/v3/client/internal/RestClient.java
@@ -26,7 +26,7 @@ import java.net.http.HttpClient;
 import java.net.http.HttpRequest;
 import java.net.http.HttpResponse;
 import java.security.SecureRandom;
-import java.util.Arrays;
+import java.util.List;
 import java.util.Map;
 import java.util.stream.Stream;
 import javax.annotation.Nonnull;
@@ -179,7 +179,7 @@ final class RestClient implements AutoCloseable {
             if (!body.isEmpty()) {
                 try {
                     final JsonNode root = objectMapper.readTree(body);
-                    for (String field : Arrays.asList("message", "error_message", "error")) {
+                    for (String field : List.of("message", "error_message", "error")) {
                         final JsonNode node = root.findValue(field);
                         if (node != null) {
                             reason = node.asText();

--- a/src/main/java/com/influxdb/v3/client/internal/RestClient.java
+++ b/src/main/java/com/influxdb/v3/client/internal/RestClient.java
@@ -26,6 +26,7 @@ import java.net.http.HttpClient;
 import java.net.http.HttpRequest;
 import java.net.http.HttpResponse;
 import java.security.SecureRandom;
+import java.util.Arrays;
 import java.util.Map;
 import java.util.stream.Stream;
 import javax.annotation.Nonnull;
@@ -35,6 +36,7 @@ import javax.net.ssl.TrustManager;
 import javax.net.ssl.X509TrustManager;
 
 import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.JsonNode;
 import com.fasterxml.jackson.databind.ObjectMapper;
 import io.netty.handler.codec.http.HttpMethod;
 import io.netty.handler.codec.http.HttpResponseStatus;
@@ -176,11 +178,13 @@ final class RestClient implements AutoCloseable {
             String body = response.body();
             if (!body.isEmpty()) {
                 try {
-                    final var obj = objectMapper.readTree(body);
-                    if (obj.hasNonNull("message")) { // Cloud
-                        reason = obj.get("message").asText();
-                    } else if (obj.hasNonNull("error_message")) { // OSS (Edge)
-                        reason = obj.get("error_message").asText();
+                    final JsonNode root = objectMapper.readTree(body);
+                    for (String field : Arrays.asList("message", "error_message", "error")) {
+                        final JsonNode node = root.findValue(field);
+                        if (node != null) {
+                            reason = node.asText();
+                            break;
+                        }
                     }
                 } catch (JsonProcessingException e) {
                     LOG.debug("Can't parse msg from response {}", response);

--- a/src/main/java/com/influxdb/v3/client/internal/RestClient.java
+++ b/src/main/java/com/influxdb/v3/client/internal/RestClient.java
@@ -179,7 +179,8 @@ final class RestClient implements AutoCloseable {
             if (!body.isEmpty()) {
                 try {
                     final JsonNode root = objectMapper.readTree(body);
-                    for (String field : List.of("message", "error_message", "error")) {
+                    final List<String> possibilities = List.of("message", "error_message", "error");
+                    for (final String field : possibilities) {
                         final JsonNode node = root.findValue(field);
                         if (node != null) {
                             reason = node.asText();

--- a/src/test/java/com/influxdb/v3/client/internal/RestClientTest.java
+++ b/src/test/java/com/influxdb/v3/client/internal/RestClientTest.java
@@ -359,7 +359,7 @@ public class RestClientTest extends AbstractMockServerTest {
     @Test
     public void errorFromBodyEdgeWithMessage() { // OSS/Edge specific error message
         mockServer.enqueue(createResponse(400)
-                .setBody("{\"error\":\"parsing failed\",\"data\":{\"error_message\":\"invalid field value in line protocol\"}}"));
+                .setBody("{\"error\":\"parsing failed\",\"data\":{\"error_message\":\"invalid field value\"}}"));
 
         restClient = new RestClient(new ClientConfig.Builder()
                 .host(baseURL)
@@ -368,7 +368,7 @@ public class RestClientTest extends AbstractMockServerTest {
         Assertions.assertThatThrownBy(
                         () -> restClient.request("ping", HttpMethod.GET, null, null, null))
                 .isInstanceOf(InfluxDBApiException.class)
-                .hasMessage("HTTP status code: 400; Message: invalid field value in line protocol");
+                .hasMessage("HTTP status code: 400; Message: invalid field value");
     }
 
     @Test

--- a/src/test/java/com/influxdb/v3/client/internal/RestClientTest.java
+++ b/src/test/java/com/influxdb/v3/client/internal/RestClientTest.java
@@ -342,6 +342,36 @@ public class RestClientTest extends AbstractMockServerTest {
     }
 
     @Test
+    public void errorFromBodyEdgeWithoutMessage() { // OSS/Edge error message
+        mockServer.enqueue(createResponse(400)
+                .setBody("{\"error\":\"parsing failed\"}"));
+
+        restClient = new RestClient(new ClientConfig.Builder()
+                .host(baseURL)
+                .build());
+
+        Assertions.assertThatThrownBy(
+                        () -> restClient.request("ping", HttpMethod.GET, null, null, null))
+                .isInstanceOf(InfluxDBApiException.class)
+                .hasMessage("HTTP status code: 400; Message: parsing failed");
+    }
+
+    @Test
+    public void errorFromBodyEdgeWithMessage() { // OSS/Edge specific error message
+        mockServer.enqueue(createResponse(400)
+                .setBody("{\"error\":\"parsing failed\",\"data\":{\"error_message\":\"invalid field value in line protocol\"}}"));
+
+        restClient = new RestClient(new ClientConfig.Builder()
+                .host(baseURL)
+                .build());
+
+        Assertions.assertThatThrownBy(
+                        () -> restClient.request("ping", HttpMethod.GET, null, null, null))
+                .isInstanceOf(InfluxDBApiException.class)
+                .hasMessage("HTTP status code: 400; Message: invalid field value in line protocol");
+    }
+
+    @Test
     public void errorFromBodyText() {
         mockServer.enqueue(createResponse(402)
                 .setBody("token is over the limit"));

--- a/src/test/java/com/influxdb/v3/client/write/WriteOptionsTest.java
+++ b/src/test/java/com/influxdb/v3/client/write/WriteOptionsTest.java
@@ -45,6 +45,21 @@ class WriteOptionsTest {
 
 
     @Test
+    void optionsSelf() {
+        WriteOptions options = new WriteOptions("my-database", WritePrecision.S, 512);
+
+        Assertions.assertThat(options).isEqualTo(options);
+    }
+
+    @Test
+    void optionsAlien() {
+        WriteOptions options = new WriteOptions("my-database", WritePrecision.S, 512);
+
+        Assertions.assertThat(options).isNotEqualTo(null);
+        Assertions.assertThat(options).isNotEqualTo(this);
+    }
+
+    @Test
     void optionsEqualAll() {
         WriteOptions options = new WriteOptions("my-database", WritePrecision.S, 512);
         WriteOptions optionsViaBuilder = new WriteOptions.Builder()

--- a/src/test/java/com/influxdb/v3/client/write/WriteOptionsTest.java
+++ b/src/test/java/com/influxdb/v3/client/write/WriteOptionsTest.java
@@ -45,16 +45,10 @@ class WriteOptionsTest {
 
 
     @Test
-    void optionsSelf() {
+    void optionsBasics() {
         WriteOptions options = new WriteOptions("my-database", WritePrecision.S, 512);
 
         Assertions.assertThat(options).isEqualTo(options);
-    }
-
-    @Test
-    void optionsAlien() {
-        WriteOptions options = new WriteOptions("my-database", WritePrecision.S, 512);
-
         Assertions.assertThat(options).isNotEqualTo(null);
         Assertions.assertThat(options).isNotEqualTo(this);
     }

